### ISSUE TITLE
gateloop_use_heinsen=True on MeshTransformer results in NaN loss

### ISSUE THREAD

**Issue by Kurokabe** [lucidrains/meshgpt-pytorch#47]
The loss quickly became NaN when training on ShapeNet after filtering for mesh < 800 faces after decimation (resulting in ~15k different 3D models) with `condition_on_text=True` so I thought it was similar to #44 but even after training without text, adjusting the learning rate, adding warmup, larger batch size...  I still had this problem.

I found with [detect_anomaly](https://pytorch.org/docs/stable/autograd.html#anomaly-detection) that the NaN come from [`gateloop_transformer>heinsen_associative_scan`](https://github.com/lucidrains/gateloop-transformer/blob/0a3aac85dee50c8e3a7c1563fad0eae0144f0414/gateloop_transformer/simplified_gate_loop.py#L30C66-L30C66) on the log backward.

Previously I was able to train successfully on 5 ShapeNet categories, with 10 meshes x 256 transformations each = 12'800 3D models, but it was with meshgpt-pytorch version 0.4.2. Maybe updating to version 0.5.5 also updated gateloop-transformer which added this bug. Anyway, setting `gateloop_use_heinsen=False` seems to have solved the problem for me.

**Comment by Kurokabe** [comment]
With the default values it's converging, but I didn't try with `coarse_pre_gateloop_depth=0` and `fine_pre_gateloop_depth=0` to compare. If I find the time to run without GateLoop I'll report here

**Comment by lucidrains** [comment]
@Kurokabe please do, i'm curious to know

**Comment by Kurokabe** [comment]
> @Kurokabe does the use of gateloop layers help with convergence?

So I tried with `coarse_pre_gateloop_depth=0` and `fine_pre_gateloop_depth=0`, I didn't train it as long as my previous model, but using the Gateloop seems to improve a bit the performance
![image](https://github.com/lucidrains/meshgpt-pytorch/assets/37635308/8015af3d-c965-4038-b810-2b93f2992c7c)

![image](https://github.com/lucidrains/meshgpt-pytorch/assets/37635308/cdc7383e-6a01-4c38-adf4-5b02caacf419)

Also, in your coarse gateloop and fine gateloop call, you pass the `cache` parameter, but in the case of `coarse_pre_gateloop_depth=0` and `fine_pre_gateloop_depth=0` it won't work since the nn.Identity() doesn't have a `cache` parameter
https://github.com/lucidrains/meshgpt-pytorch/blob/5dc77d9b3915558bdb5f3edacbb22468fe27848c/meshgpt_pytorch/meshgpt_pytorch.py#L1483

I temporarily fixed it as follows:
```python
if isinstance(self.coarse_gateloop_block, GateLoopBlock):
    face_codes, coarse_gateloop_cache = self.coarse_gateloop_block(face_codes, cache = coarse_gateloop_cache)
else:
    coarse_gateloop_cache = None
    face_codes = self.coarse_gateloop_block(face_codes)
```

**Comment by lucidrains** [comment]
@Kurokabe ah nice, thank you for sharing that!

also fixed the issue with being able to run without gateloop layers
